src/frontend/src/utils/countdown.ts

```diff
@@ -0,0 +1,78 @@
+import { render } from "lit-html";
+import { displayError } from "../components/displayError";
+
+export class Countdown {
+  private timeoutHandle;
+  private expected;
+  private readonly endTimestamp;
+
+  constructor(
+    private updateFunc: () => void,
+    private interval: number,
+    private timeoutFunc: () => void,
+    endTimestamp: bigint
+  ) {
+    this.expected = Date.now() + this.interval;
+    this.timeoutHandle = window.setTimeout(() => this.step(), this.interval);
+    this.endTimestamp = Number(endTimestamp / BigInt("1000000"));
+  }
+
+  public stop(): void {
+    window.clearTimeout(this.timeoutHandle);
+  }
+
+  private async step() {
+    const now = Date.now();
+    if (now >= this.endTimestamp) {
+      await this.timeoutFunc();
+      return;
+    }
+    const drift = now - this.expected;
+    this.updateFunc();
+    this.expected += this.interval;
+    this.timeoutHandle = window.setTimeout(
+      () => this.step(),
+      Math.max(0, this.interval - drift)
+    );
+  }
+}
+
+export const setupCountdown = (
+  timestamp: bigint,
+  continueFunc: () => Promise<void>
+): Countdown => {
+  const timer = document.getElementById("timer") as HTMLElement;
+  return new Countdown(
+    () => render(formatRemainingTime(timestamp), timer),
+    1000,
+    async () => {
+      await displayError({
+        title: "Timeout Reached",
+        message:
+          'The timeout has been reached. For security reasons the "add device" process has been aborted.',
+        primaryButton: "Ok",
+      });
```

Comment:
The behavior on "error" should probably be customizable; it's really odd having a "util" module that references a feature like "add device" directly